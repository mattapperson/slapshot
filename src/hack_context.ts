expect.extend({
  __slapshot__hack__context(_received, fn) {
    fn({
      // @ts-ignore
      assertionCalls: this.assertionCalls,
      // @ts-ignore
      currentTestName: this.currentTestName,
      // @ts-ignore
      testPath: this.testPath
    });
    return { pass: true, message: () => "expected the slapshot hack to work" };
  }
});
