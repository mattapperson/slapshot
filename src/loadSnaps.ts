import fs from "fs";
import vm from "vm";
export function loadSnaps(snapshotPath: string) {
  if (!fs.existsSync(snapshotPath)) {
    return {};
  }
  const sandbox = {
    exports: {}
  };
  const source = fs.readFileSync(snapshotPath, "utf8");
  try {
    vm.runInNewContext(source, sandbox);
    return sandbox.exports;
  } catch (e) {
    console.error("Could not load file", snapshotPath);
    console.error(source);
    console.error(e);
    return {};
  }
}
