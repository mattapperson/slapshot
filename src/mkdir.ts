import path from "path";
import fs from "fs";
export const mkdir = (filePath: string) => {
  var dir = path.dirname(filePath);
  if (fs.existsSync(dir)) return true;
  mkdir(dir);
  fs.mkdirSync(dir);
};
