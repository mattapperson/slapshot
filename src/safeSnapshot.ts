export function safeSnapshot(obj: any, toSnapshot: boolean = true) {
  if (typeof obj !== "object") {
    return obj;
  }
  if (obj === null) {
    return null;
  }
  return Object.keys(obj).reduce(
    (safeObject, key) => {
      if (
        !toSnapshot &&
        safeObject[key] === "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
      ) {
        safeObject[key] = () => {
          throw new Error(
            `${key} can not be called directly. It must first be mocked with or without using memorize.`
          );
        };
      }
      if (toSnapshot && typeof safeObject[key] === "function") {
        safeObject[key] = "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]";
      }
      if (typeof safeObject[key] === "object") {
        safeObject[key] = safeSnapshot(safeObject[key], toSnapshot);
      }
      return safeObject;
    },
    { ...obj }
  );
}
