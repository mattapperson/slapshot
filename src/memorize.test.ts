/* eslint-env jest */
import rimraf from "rimraf";
import { memorize } from "./memorize";
import fs from "fs";
import { performance } from "perf_hooks";
import "./hack_context";

const snapshotDir = `${__dirname}/__memorize_snapshots__`;

const fetchData = () => {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve({ foo: "bar" });
    }, 200);
  });
};

const complexReturn = (someValue: any = null) => {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(someValue || { foo: "bar", method: () => ({ here: "now" }) });
    }, 200);
  });
};

let mockPromise: any;
let mockThunk: any;

beforeEach(() => {
  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.argv = process.argv.filter(e => e !== "-u");
  process.env.SLAPSHOT_ONLINE = undefined;

  mockPromise = jest.fn(fetchData);
  mockThunk = jest.fn(mockPromise);
  rimraf.sync(snapshotDir);
});

test("calls thunk on first run", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  const data = await memorize("a", mockThunk);
  expect(mockPromise.mock.calls.length).toBe(1);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(data.foo).toBe("bar");
});

test("writes a __data_snapshot__ file to disk", async () => {
  expect(fs.existsSync(snapshotDir)).toBe(false);
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  await memorize("b", mockThunk);
  expect(fs.existsSync(snapshotDir)).toBe(true);
});

test("resolves from disk on 2nd hit", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  const liveStart = performance.now();
  await memorize("c", mockThunk);
  const liveEnd = performance.now();

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const recordedStart = performance.now();
  const data = await memorize("c", jest.fn());
  const recordedEnd = performance.now();

  expect(liveEnd - liveStart).toBeGreaterThan(190);
  expect(recordedEnd - recordedStart).toBeLessThan(10);
  expect(mockPromise.mock.calls.length).toBe(1);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(data.foo).toBe("bar");
});

test("calls run but dont update with SLAPSHOT_ONLINE", async () => {
  const mockWithValue = jest.fn(complexReturn);

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("d", mockWithValue);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  expect(() => {
    const data: any = memorize("d", () => mockWithValue({ foo: "foo" }));
    expect(mockWithValue.mock.calls.length).toBe(2);
    expect(mockWithValue.mock.calls.length).toBe(2);
    expect(data.foo).toBe("foo");
  }).toThrow();

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data2: any = await memorize("d", () => mockWithValue({ foo: "foo" }));
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(data2.foo).toBe("bar");
});

test("calls run still in unit tests with just updateSnapshot", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("d", mockThunk);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data: any = await memorize("d", mockThunk);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(data.foo).toBe("bar");
});

test("calls run and update with updateSnapshot and SLAPSHOT_ONLINE", async () => {
  const mockWithValue = jest.fn(complexReturn);

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("d", mockWithValue);

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  const data: any = await memorize("d", () => mockWithValue({ foo: "foo" }));
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(data.foo).toBe("foo");

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data2: any = await memorize("d", () => mockWithValue({ foo: "foo" }));
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(data2.foo).toBe("foo");
});

test("works with strings and numbers", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  memorize("c", () => 22);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data = memorize("c", () => {});
  expect(data).toBe(22);
});

test("nested APIs with functions throw an error when called unless mocked manualy", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("complex", complexReturn);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data: any = await memorize("complex", complexReturn);
  expect(data.foo).toBe("bar");
  expect(typeof data.foo).toBe("string");
  expect(typeof data.method).toBe("function");

  expect(() => {
    data.method();
  }).toThrow();
});

test("Impure memorized methods also add call count to name", async () => {
  const result1 = await memorize(
    "complex",
    () => {
      return 22;
    },
    { pure: false }
  );

  const result2 = await memorize(
    "complex",
    () => {
      return 21;
    },
    { pure: false }
  );

  expect(result1).toBe(22);
  expect(result2).toBe(21);
});

test("thrown errors are replayed", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  const mockedCB = jest.fn();
  expect(() => {
    memorize("error", () => {
      throw new Error("foo");
    });
  }).toThrowError("foo");

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  expect(() => {
    memorize("error", mockedCB);
  }).toThrowError("foo");

  expect(mockedCB).not.toBeCalled();
});

test("thrown errors of a custom type are replayed as the correct instance", async () => {
  class CustomError extends Error {
    constructor(message: any) {
      super(message);
      this.name = "CustomError";
    }
  }

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  const mockedCB = jest.fn();
  expect(() => {
    memorize("error", () => {
      throw new CustomError("foo");
    });
  }).toThrowError("foo");

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";

  try {
    memorize("error", mockedCB);
  } catch (e) {
    expect(e).toBeInstanceOf(CustomError);
  }
});
