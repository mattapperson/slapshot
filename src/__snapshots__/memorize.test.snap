
exports['calls thunk on first run - a'] = {
  "results": {
    "foo": "bar"
  }
}

exports['writes a __data_snapshot__ file to disk - b'] = {
  "results": {
    "foo": "bar"
  }
}

exports['resolves from disk on 2nd hit - c'] = {
  "results": {
    "foo": "bar"
  }
}

exports['calls run but dont update with SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['calls run but dont update with SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "foo"
  }
}

exports['calls run still in unit tests with just updateSnapshot - d'] = {
  "results": {
    "foo": "bar"
  }
}

exports['calls run and update with updateSnapshot and SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['calls run and update with updateSnapshot and SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "foo"
  }
}

exports['works with strings and numbers - c'] = {
  "results": 22,
  "thrownError": null
}

exports['works with null value - c'] = {
  "results": null,
  "thrownError": null
}

exports['works with JSON string in memorize name - JSON - {"foo":"dog"}'] = {
  "results": {
    "foo": "bar"
  }
}

exports['works with undefined value - c'] = {
  "thrownError": null
}

exports['works with array value - c'] = {
  "results": [
    1,
    2,
    3
  ],
  "thrownError": null
}

exports['nested APIs with functions throw an error when called unless mocked manualy - complex'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['No race conditions - works many times out of sync - c - 1'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 2'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 3'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 4'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 5'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 6'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 7'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 8'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 9'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 10'] = {
  "results": 22
}

exports['thrown errors are replayed - error'] = {
  "results": null,
  "thrownError": "{\"type\":\"Error\",\"message\":\"foo\",\"name\":\"Error\",\"stack\":\"Error: foo\\n    at memorize_1.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:386:27)\\n    at Object.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.js:52:46)\\n    at expect (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:382:38)\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\n    at /home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:389:12\\n    at Generator.next (<anonymous>)\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\n    at new Promise (<anonymous>)\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\n    at Object.<anonymous>.test (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:375:34)\\n    at Object.e.length.arguments.(anonymous function) (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/wallaby/runners/node/jest@0.4.3/jasmine2Initializer.js:14:1786)\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\n    at new Promise (<anonymous>)\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\"}"
}

exports['thrown errors are replayed with custom properties - error'] = {
  "results": null,
  "thrownError": "{\"type\":\"Error\",\"message\":\"foo\",\"name\":\"Error\",\"stack\":\"Error: foo\\n    at memorize_1.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:421:27)\\n    at Object.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.js:52:46)\\n    at expect (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:417:38)\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\n    at /home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:424:12\\n    at Generator.next (<anonymous>)\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\n    at new Promise (<anonymous>)\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\n    at Object.<anonymous>.test (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:404:34)\\n    at Object.e.length.arguments.(anonymous function) (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/wallaby/runners/node/jest@0.4.3/jasmine2Initializer.js:14:1786)\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\n    at new Promise (<anonymous>)\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\",\"customProp\":\"bar\"}"
}

exports['consoles error of non-matching snap when validateSnapshot is not set - c'] = {
  "results": 22,
  "thrownError": null
}

exports['throws error of non-matching snap when validateSnapshot is set to true - validateSnapshot'] = {
  "results": 22,
  "thrownError": null
}

exports['deffers to validateSnapshot to validate snapshot when validateSnapshot is a funtion - c'] = {
  "results": 22,
  "thrownError": null
}

exports['calls thunk on first run - a'] = {
  "results": {
    "foo": "bar"
  }
}

exports['writes a __data_snapshot__ file to disk - b'] = {
  "results": {
    "foo": "bar"
  }
}

exports['resolves from disk on 2nd hit - c'] = {
  "results": {
    "foo": "bar"
  }
}

exports['calls run but dont update with SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['calls run but dont update with SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "foo"
  }
}

exports['calls run still in unit tests with just updateSnapshot - d'] = {
  "results": {
    "foo": "bar"
  }
}

exports['calls run and update with updateSnapshot and SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['calls run and update with updateSnapshot and SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "foo"
  }
}

exports['works with strings and numbers - c'] = {
  "results": 22,
  "thrownError": null
}

exports['works with null value - c'] = {
  "results": null,
  "thrownError": null
}

exports['works with JSON string in memorize name - JSON - {"foo":"dog"}'] = {
  "results": {
    "foo": "bar"
  }
}

exports['works with undefined value - c'] = {
  "thrownError": null
}

exports['works with array value - c'] = {
  "results": [
    1,
    2,
    3
  ],
  "thrownError": null
}

exports['nested APIs with functions throw an error when called unless mocked manualy - complex'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['No race conditions - works many times out of sync - c - 1'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 2'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 3'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 4'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 5'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 6'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 7'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 8'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 9'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 10'] = {
  "results": 22
}

exports['thrown errors are replayed - error'] = {
  "results": null,
  "thrownError": "{\"type\":\"Error\",\"message\":\"foo\",\"name\":\"Error\",\"stack\":\"Error: foo\\n    at memorize_1.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:386:27)\\n    at Object.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.js:52:46)\\n    at expect (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:382:38)\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\n    at /home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:389:12\\n    at Generator.next (<anonymous>)\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\n    at new Promise (<anonymous>)\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\n    at Object.<anonymous>.test (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:375:34)\\n    at Object.e.length.arguments.(anonymous function) (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/wallaby/runners/node/jest@0.4.3/jasmine2Initializer.js:14:1786)\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\n    at new Promise (<anonymous>)\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\"}"
}

exports['thrown errors are replayed with custom properties - error'] = {
  "results": null,
  "thrownError": "{\"type\":\"Error\",\"message\":\"foo\",\"name\":\"Error\",\"stack\":\"Error: foo\\n    at memorize_1.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:421:27)\\n    at Object.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.js:52:46)\\n    at expect (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:417:38)\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\n    at /home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:424:12\\n    at Generator.next (<anonymous>)\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\n    at new Promise (<anonymous>)\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\n    at Object.<anonymous>.test (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:404:34)\\n    at Object.e.length.arguments.(anonymous function) (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/wallaby/runners/node/jest@0.4.3/jasmine2Initializer.js:14:1786)\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\n    at new Promise (<anonymous>)\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\",\"customProp\":\"bar\"}"
}

exports['consoles error of non-matching snap when validateSnapshot is not set - c'] = {
  "results": 22,
  "thrownError": null
}

exports['throws error of non-matching snap when validateSnapshot is set to true - validateSnapshot'] = {
  "results": 22,
  "thrownError": null
}

exports['deffers to validateSnapshot to validate snapshot when validateSnapshot is a funtion - c'] = {
  "results": 22,
  "thrownError": null
}

exports['calls thunk on first run - a'] = {
  "results": {
    "foo": "bar"
  }
}

exports['writes a __snapshot__ file to disk - b'] = {
  "results": {
    "foo": "bar"
  }
}

exports['resolves from disk on 2nd hit - c'] = {
  "results": {
    "foo": "bar"
  }
}

exports['calls thunk on first run - a'] = {
  "results": {
    "foo": "bar"
  }
}

exports['writes a __snapshot__ file to disk - b'] = {
  "results": {
    "foo": "bar"
  }
}

exports['resolves from disk on 2nd hit - c'] = {
  "results": {
    "foo": "bar"
  }
}

exports['calls run but dont update with SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['calls run but dont update with SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "foo"
  }
}

exports['calls run still in unit tests with just updateSnapshot - d'] = {
  "results": {
    "foo": "bar"
  }
}

exports['calls run and update with updateSnapshot and SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['calls run and update with updateSnapshot and SLAPSHOT_ONLINE - d'] = {
  "results": {
    "foo": "foo"
  }
}

exports['works with strings and numbers - c'] = {
  "results": 22,
  "thrownError": null
}

exports['works with null value - c'] = {
  "results": null,
  "thrownError": null
}

exports['works with JSON string in memorize name - JSON - {"foo":"dog"}'] = {
  "results": {
    "foo": "bar"
  }
}

exports['works with undefined value - c'] = {
  "thrownError": null
}

exports['works with array value - c'] = {
  "results": [
    1,
    2,
    3
  ],
  "thrownError": null
}

exports['nested APIs with functions throw an error when called unless mocked manualy - complex'] = {
  "results": {
    "foo": "bar",
    "method": "[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"
  }
}

exports['No race conditions - works many times out of sync - c - 1'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 2'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 3'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 4'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 5'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 6'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 7'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 8'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 9'] = {
  "results": 22
}

exports['No race conditions - works many times out of sync - c - 10'] = {
  "results": 22
}

exports['thrown errors are replayed - error'] = {
  "results": null,
  "thrownError": "{\"type\":\"Error\",\"message\":\"foo\",\"name\":\"Error\",\"stack\":\"Error: foo\\n    at memorize_1.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:386:27)\\n    at Object.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.js:52:46)\\n    at expect (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:382:38)\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\n    at /home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:389:12\\n    at Generator.next (<anonymous>)\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\n    at new Promise (<anonymous>)\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\n    at Object.<anonymous>.test (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:375:34)\\n    at Object.e.length.arguments.(anonymous function) (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/wallaby/runners/node/jest@0.4.3/jasmine2Initializer.js:14:1786)\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\n    at new Promise (<anonymous>)\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\"}"
}

exports['thrown errors are replayed with custom properties - error'] = {
  "results": null,
  "thrownError": "{\"type\":\"Error\",\"message\":\"foo\",\"name\":\"Error\",\"stack\":\"Error: foo\\n    at memorize_1.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:421:27)\\n    at Object.memorize (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.js:52:46)\\n    at expect (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:417:38)\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\n    at /home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:424:12\\n    at Generator.next (<anonymous>)\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\n    at new Promise (<anonymous>)\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\n    at Object.<anonymous>.test (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/projects/f5531a1be1fa0655/instrumented/src/memorize.test.js:404:34)\\n    at Object.e.length.arguments.(anonymous function) (/home/coder/project/data/extensions/wallabyjs.wallaby-vscode-1.0.147/wallaby/runners/node/jest@0.4.3/jasmine2Initializer.js:14:1786)\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\n    at new Promise (<anonymous>)\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\",\"customProp\":\"bar\"}"
}

exports['consoles error of non-matching snap when validateSnapshot is not set - c'] = {
  "results": 22,
  "thrownError": null
}

exports['throws error of non-matching snap when validateSnapshot is set to true - validateSnapshot'] = {
  "results": 22,
  "thrownError": null
}

exports['deffers to validateSnapshot to validate snapshot when validateSnapshot is a funtion - c'] = {
  "results": 22,
  "thrownError": null
}
